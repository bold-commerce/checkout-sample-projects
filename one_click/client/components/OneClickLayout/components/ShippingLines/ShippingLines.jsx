/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import RadioField from '@boldcommerce/stacks-ui/lib/components/radiofield/RadioField';
import { Price } from '@boldcommerce/stacks-ui';
import EmptyState from '../EmptyState/EmptyState';
import LoadingState from '../LoadingState/LoadingState';
import { useCheckoutStore, useShippingLines } from '@boldcommerce/checkout-react-components';
import './ShippingLines.css';

export const ShippingLines = ({
  showShippingLines, shippingLinesFetching, shippingLinesLoadingStatus, shippingLines, selectedShippingLineIndex, setSelectedShippingLine,
}) => {
  if (!showShippingLines) {
    return (
      <section className="FieldSet FieldSet--ShippingMethod">
        <div className="FieldSet__Header">
          <h3 className="FieldSet__Heading">Shipping method</h3>
        </div>
        <div className="FieldSet__Content"><EmptyState title="To view shipping options, complete filling in your address" /></div>
      </section>
    );
  }

  return (
    <section className="FieldSet FieldSet--ShippingMethod">
      <div className="FieldSet__Header">
        <h3 className="FieldSet__Heading">Shipping method</h3>
      </div>
      {
        shippingLinesFetching ? <LoadingState />
          : (
            <div className="FieldSet__Content">
              {shippingLines && shippingLines.map((method, index) => (
                <div className="RadioButton" key={index}>
                  <RadioField
                    label={method.description}
                    name="shipping-method"
                    checked={selectedShippingLineIndex === parseInt(method.id, 10)}
                    className="RadioField"
                    disabled={shippingLinesLoadingStatus === 'setting'}
                    onChange={() => setSelectedShippingLine(index)}
                  />
                  <Price className="ShippingMethod__Price" amount={method.amount} />
                </div>
              ))}
            </div>
          )
      }
    </section>
  );
};

ShippingLines.propTypes = {
  showShippingLines: PropTypes.bool,
  shippingLinesFetching: PropTypes.bool,
  shippingLinesLoadingStatus: PropTypes.string,
  shippingLines: PropTypes.array,
  selectedShippingLineIndex: PropTypes.number,
  setSelectedShippingLine: PropTypes.func,
};

const MemoizedShippingLines = React.memo(ShippingLines);

const ShippingLinesContainer = () => {
  const { data, loadingStatus, errors, updateShippingLine,  } = useShippingLines();
  const { shippingLines, selectedShippingLineIndex } = data;
  const { state } = useCheckoutStore();
  const country_code = state?.applicationState?.addresses?.shipping?.country_code;

  return (
    <MemoizedShippingLines
      showShippingLines={ country_code && loadingStatus !== 'incomplete' && !errors }
      shippingLinesFetching={ loadingStatus === 'fetching' }
      shippingLinesLoadingStatus={loadingStatus}
      shippingLines={shippingLines}
      selectedShippingLineIndex={selectedShippingLineIndex}
      setSelectedShippingLine={updateShippingLine}
    />
  );
};

export default ShippingLinesContainer;
