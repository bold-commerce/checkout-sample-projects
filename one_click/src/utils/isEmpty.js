const isEmpty = (obj) => {
  return JSON.stringify(obj) === JSON.stringify({});
};

export default isEmpty;