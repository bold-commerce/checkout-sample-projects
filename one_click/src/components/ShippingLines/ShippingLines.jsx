/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { useCheckoutStore, useShippingLines } from '@boldcommerce/checkout-react-components';
import { RadioField, Price } from '@boldcommerce/stacks-ui';
import LoadingState from '../LoadingState/LoadingState';
import { useAnalytics, useErrorLogging } from '../../hooks';
import './ShippingLines.css';
import { useTranslation } from 'react-i18next';

export const ShippingLines = ({
  showShippingLines, shippingLinesFetching, shippingLinesLoadingStatus, shippingLines, selectedShippingLineIndex, setSelectedShippingLine, disabled
}) => {
  const trackEvent = useAnalytics();
  const logError = useErrorLogging();
  const { t } = useTranslation();

  const handleShippingLineChange = async (index) => {
    try {
      await setSelectedShippingLine(index);
      trackEvent('set_shipping_line');
    } catch(e) {
      logError('shipping_line', e);
    }
  };

  if (!showShippingLines) {
    return (
      <section className="FieldSet FieldSet--ShippingMethod">
        <div className="FieldSet__Header">
          <h3 className="FieldSet__Heading">{t('shipping.method')}</h3>
        </div>
        <div className="FieldSet__Content">
          <p>{t('shipping.options_description')}</p>
        </div>
      </section>
    );
  }

  return (
    <section className="FieldSet FieldSet--ShippingMethod">
      <div className="FieldSet__Header">
        <h3 className="FieldSet__Heading">{t('shipping.method')}</h3>
      </div>
      {
        shippingLinesFetching ? <LoadingState />
          : (
            <div className="FieldSet__Content">
              {shippingLines && shippingLines.map((method, index) => (
                <div className="RadioButton" key={index}>
                  <RadioField
                    label={method.description}
                    name="shipping-method"
                    checked={selectedShippingLineIndex === parseInt(method.id, 10)}
                    className="RadioField"
                    disabled={shippingLinesLoadingStatus === 'setting' || disabled }
                    onChange={() => handleShippingLineChange(index)}
                  />
                  <Price className="ShippingMethod__Price" amount={method.amount} />
                </div>
              ))}
            </div>
          )
      }
    </section>
  );
};

ShippingLines.propTypes = {
  showShippingLines: PropTypes.bool,
  shippingLinesFetching: PropTypes.bool,
  shippingLinesLoadingStatus: PropTypes.string,
  shippingLines: PropTypes.array,
  selectedShippingLineIndex: PropTypes.number,
  setSelectedShippingLine: PropTypes.func,
};

const MemoizedShippingLines = React.memo(ShippingLines);

const ShippingLinesContainer = ( {activePage, disabled} ) => {
  const { data, loadingStatus, errors, updateShippingLine,  } = useShippingLines();
  const { shippingLines, selectedShippingLineIndex } = data;
  const { state } = useCheckoutStore();
  const country_code = state?.applicationState?.addresses?.shipping?.country_code;

  return (
    <MemoizedShippingLines
      showShippingLines={ country_code && loadingStatus !== 'incomplete' && !errors && activePage }
      shippingLinesFetching={ loadingStatus === 'fetching' }
      shippingLinesLoadingStatus={loadingStatus}
      shippingLines={shippingLines}
      selectedShippingLineIndex={selectedShippingLineIndex}
      setSelectedShippingLine={updateShippingLine}
      disabled={disabled}
    />
  );
};

export default ShippingLinesContainer;
