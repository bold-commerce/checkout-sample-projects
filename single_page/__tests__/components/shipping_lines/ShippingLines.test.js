import React from "react";
import { render } from "@testing-library/react";
import { exampleLineItems as MOCKlineItems } from '../../utils/lineItemHelpers';
import { 
  exampleBreakdown as MOCKbreakdown,
  exampleShippingLines as MOCKshippingLines
} from '../../utils/orderHelpers';
import {
  countries as MOCKcountries,
  caProvinces as MOCKprovinces,
  exampleAddress as MOCKaddress
} from '../../utils/addressHelpers';
import ShippingLines from "../../../client/components/ShippingLines/ShippingLines";

jest.mock('@boldcommerce/checkout-react-components', () => ({
  useBillingAddress: () => ({
    data: {MOCKaddress},
    submitBillingAddress: (() => {})
  }),
  useBillingSameAsShipping: () => ({
    setBillingSameAsShipping: (() => {}),
    data: true
  }),
  useCountryInfo: () => ({
    data: {
      countries: MOCKcountries,
      provinces: MOCKprovinces,
      provinceLabel: "province",
      showPostalCode: true,
      showProvince: true
    }
  }),
  usePaymentIframe : () => ({
    data: {
      url: 'test.gg',
      height: 42
    },
    processPaymentIframe: (() => {}),
    paymentIframeOnLoaded: (() => {})
  }),
  useCheckoutStore: () => ({
    state: {
      orderInfo: 'fulfilled',
      loadingStatus: {
        isLoading: false
      }
    }
  }),
  useLineItems: () => ({ 
    data : MOCKlineItems
  }),
  useBreakdown: () => ({
    data: MOCKbreakdown
  }),
  useDiscount: () => ({
    data: {
      discountApplied: true,
      discountCode: 'Test'
    },
    applyDiscount: (() => {})
  }),
  useShippingAddress: () => ({
    data: MOCKaddress
  }),
  useShippingLines: () => ({
    data: {
      shippingLines: MOCKshippingLines
    },
    updateShippingLines: (() => {}),
    getShippingLines: (() => { return MOCKshippingLines })
  }),
  useLoadingStatus: () => ({
    data: {
      shippingAddress: 'fulfilled'
    }
  }),
  useErrors: () => ({
    data: { shippingAddress: false}
  })
}))

describe('ShippingLines', () => {
  test('renders ShippingLines component', () => {
    const { asFragment } = render(<ShippingLines/>)

    expect(asFragment()).toMatchSnapshot();
  })
})
