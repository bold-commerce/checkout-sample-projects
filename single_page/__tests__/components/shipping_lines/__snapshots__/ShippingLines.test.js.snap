// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShippingLines renders ShippingLines component 1`] = `
<DocumentFragment>
  <section
    class="FieldSet FieldSet--ShippingMethod"
  >
    <div
      class="FieldSet__Header"
    >
      <h2
        class="FieldSet__Heading"
      >
        Shipping method
      </h2>
    </div>
    <div
      class="FieldSet__Content"
    >
      <div
        class="RadioButton"
      >
        <div
          class="stx-field stx-field--with-radio-field RadioField"
        >
          <div
            class="stx-radio-field"
          >
            <span
              class="stx-radio stx-radio-field__radio"
            >
              <input
                aria-checked="false"
                aria-disabled="false"
                class="stx-radio__input"
                id="stxField1"
                name="shipping-method"
                role="radio"
                type="radio"
                value=""
              />
              <span
                class="stx-radio__box"
              />
            </span>
            <label
              class="stx-radio-field__label"
              for="stxField1"
            >
              Standard Shipping
            </label>
          </div>
        </div>
        <div
          class="stx-price ShippingMethod__Price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            $20.00
          </p>
        </div>
      </div>
      <div
        class="RadioButton"
      >
        <div
          class="stx-field stx-field--with-radio-field RadioField"
        >
          <div
            class="stx-radio-field"
          >
            <span
              class="stx-radio stx-radio-field__radio"
            >
              <input
                aria-checked="false"
                aria-disabled="false"
                class="stx-radio__input"
                id="stxField2"
                name="shipping-method"
                role="radio"
                type="radio"
                value=""
              />
              <span
                class="stx-radio__box"
              />
            </span>
            <label
              class="stx-radio-field__label"
              for="stxField2"
            >
              Expedited Shipping
            </label>
          </div>
        </div>
        <div
          class="stx-price ShippingMethod__Price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            $40.00
          </p>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
